src/cluster.py

```diff
@@ -0,0 +1,212 @@
+        except (subprocess.CalledProcessError, subprocess.TimeoutExpired) as e:
+            logger.warning(e.stderr)
+            raise e
+
+
+class ClusterUpgrades(ops.framework.Object):
+    """ClusterUpgrades manages snap upgrades across the cluster."""
+
+    charm = None
+    _stored = ops.framework.StoredState()
+
+    def __init__(self, charm: "charm.MicroCephCharm"):
+        super().__init__(charm, "cluster-upgrade")
+        self.charm = charm
+
+    @property
+    def peer_int(self):
+        """Get the peer interface."""
+        return self.charm.peers.interface
+
+    @property
+    def channel(self) -> str:
+        """Get the snap channel."""
+        return self.charm.channel
+
+    @channel.setter
+    def channel(self, value: str) -> None:
+        self.charm.channel = value
+
+    @property
+    def track(self) -> str:
+        """Get the snap track."""
+        c = self.channel
+        if not self.channel:
+            c = self.model.config["snap-channel"]
+        return c.split("/")[0]
+
+    def upgrade_requested(self, chan: str) -> bool:
+        """Check if a snap upgrade was requested."""
+        logger.debug(f"Requested, current channel: {chan}, {self.channel}")
+        return self.channel != chan
+
+    def can_upgrade_snap(self, snap_chan: str) -> Tuple[bool, str]:
+        """Check if snap can be upgraded."""
+        if not microceph.can_upgrade_snap(self.track, snap_chan.split("/")[0]):
+            msg = f"Cannot upgrade from {self.channel} to {snap_chan}"
+            logger.warning(msg)
+            return False, msg
+        health, det = CephStatus().ceph_health()
+        if health != CephHealth.Ok:
+            msg = f"Cannot upgrade, ceph health not ok: {health}, {det}"
+            logger.warning(msg)
+            return False, msg
+        return True, ""
+
+    def perform_upgrade(self, channel: str) -> None:
+        """Perform the snap upgrade on this node."""
+        node = self.model.unit.name
+        logger.debug(f"Upgrading {node} to {channel}")
+
+        # Check if any of the non-service commands are running
+        # Upgrading while a non-service command is running fails; checking
+        # this here so we can return a meaningful error message
+        cmds_pat = r"/snap/microceph/.*/(microceph|ceph|rados|rbd)"
+        try:
+            # As some of these are Python scripts we need to check against the
+            # full command line
+            subprocess.run(["pgrep", "-f", cmds_pat], check=True)
+            msg = "Cannot upgrade, one of microceph|ceph|rados|rbd commands is running"
+            logger.warning(msg)
+            raise sunbeam_guard.BlockedExceptionError(msg)
+        except subprocess.CalledProcessError:
+            # pgrep didn't find the command and returned non-zero
+            logger.debug("check running programs: none running")
+
+        # TODO(peter) possibly set noout, noin for cases where upgrades take longer
```

Comment:
nit: unresolved TODO

Reply:
Just a reminder for a future traveller on a possible improvement

Reply:
ack